planner/cascades/transformation_rules.go

```diff
@@ -963,33 +966,32 @@ type EliminateProjection struct {
 // The pattern of this rule is `Projection -> Any`.
 func NewRuleEliminateProjection() Transformation {
 	rule := &EliminateProjection{}
-	rule.pattern = memo.BuildPattern(
-		memo.OperandProjection,
-		memo.EngineTiDBOnly,
-		memo.NewPattern(memo.OperandAny, memo.EngineTiDBOnly),
-	)
+	rule.pattern = memo.NewPattern(memo.OperandProjection, memo.EngineTiDBOnly)
 	return rule
 }
 
 // OnTransform implements Transformation interface.
 // This rule tries to eliminate the projection whose output columns are the same with its child.
 func (r *EliminateProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
-	child := old.Children[0]
-	if child.Group.Prop.Schema.Len() != old.GetExpr().Group.Prop.Schema.Len() {
+	child := old.GetExpr().Children[0]
+	if child.Prop.Schema.Len() != old.GetExpr().Group.Prop.Schema.Len() {
 		return nil, false, false, nil
 	}
 
 	oldCols := old.GetExpr().Group.Prop.Schema.Columns
-	for i, col := range child.Group.Prop.Schema.Columns {
+	for i, col := range child.Prop.Schema.Columns {
 		if !col.Equal(nil, oldCols[i]) {
 			return nil, false, false, nil
 		}
 	}
 
 	// Promote the children group's expression.
-	finalGroupExprs := make([]*memo.GroupExpr, 0, child.Group.Equivalents.Len())
-	for elem := child.Group.Equivalents.Front(); elem != nil; elem = elem.Next() {
-		finalGroupExprs = append(finalGroupExprs, elem.Value.(*memo.GroupExpr))
```

Comment:
It is not safe to reuse a `GroupExpr`. So we need to create a new `GroupExpr` with the same `LogicalPlan`.